time integrated flux diagnostics

rgknox opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description:

This adds a site-level diagnostic that compares the total amount of mass in live vegetation (plants plus seeds) to the total integrated fluxes in and out of that pool over the life of the simulation.  The same is true for the litter pool. History output for the bias on this comparison is added as well, to see if the bias grows over time.

<!--- Describe your changes in detail -->
<!--- please add issue number if one exists -->

### Collaborators:

@JessicaNeedham 

<!--- List names of collaborators or people who have interacted -->
<!--- in bringing about this set of changes -->
<!--- consultation, discussions, etc. -->

### Expectation of Answer Changes:

This **does** have a bug fix, so some diagnostics related to litter fluxes may be corrected.

<!--- Please describe under what conditions, if any, -->
<!--- the model is expected to generated different answers -->
<!--- from the master version of the code -->

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
*If this is your first time contributing, please read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/main/CONTRIBUTING.md) document.*

All checklist items must be checked to enable merging this pull request:

*Contributor*
- [ ] The in-code documentation has been updated with descriptive comments
- [ ] The documentation has been assessed to determine if updates are necessary

*Integrator*
- [ ] FATES PASS/FAIL regression tests were run
- [ ] Evaluation of test results for answer changes was performed and results provided

### Documentation
<!--- If this pull requests warrants an update to the tech doc or user's guide, and said changes have been made paste a link to the documentation pull request below.  -->
<!--- If documentation updates are needed, but changes do not yet have their own separate pull request, please create an issue on either repo so that we can keep track of necessary updates.-->
- [Technical Note](https://github.com/NGEET/fates-docs) update:
- [User's Guide](https://github.com/NGEET/fates-users-guide) update: 

### Test Results:
<!--- Non-trivial changes require the PASS/FAIL regression tests. -->
<!--- If changes to code are NOT expected to change answers, tests must -->
<!--- be run against a baseline. -->

*CTSM (or) E3SM (specify which) test hash-tag:*

*CTSM (or) E3SM (specify which) baseline hash-tag:*

*FATES baseline hash-tag:*

*Test Output:*

<!--- paste in test results here -->


<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

rgknox commented:
The bias between integrated flux and live-vegetation and litter states are looking pretty good now, see plots below, the scale on liveveg and litter is e-8 and e-14 respectively. 

I had to move the calculation of the flux integration to later in the call sequence so that termination mortality that occurs after vegetation dynamics is encapsulated.  I'll also run a test with CNP to see if this was the cause of #1218 .


![vegc](https://github.com/NGEET/fates/assets/5891980/c4cc6acc-bec1-4039-a9aa-28c9cfdfafe2)
![interr_litter](https://github.com/NGEET/fates/assets/5891980/f38ddf0a-6d32-400e-9b72-1a41ea2c2cf0)
![interr_liveveg](https://github.com/NGEET/fates/assets/5891980/4a2c8838-fe12-452f-b33d-00f7d36de4a5)

mpaiao reviewed (commented):
Thanks for addressing the mass conservation assessment, @rgknox ! I went through the code and it looks great, I just had a couple of minor clarification questions and suggestions.

JessicaNeedham commented on main/ChecksBalancesMod.F90:
```diff
@@ -240,7 +243,134 @@ subroutine CheckLitterPools(currentSite,bc_in)
+            case(phosphorus_element)
+               net_uptake = site_mass%net_root_uptake*area_inv
+            case default
+               write(fates_log(),*) 'FATES: Invalid choice for cohort_fusion_conservation_method'
```
Looks like the message to FATES log needs updating. 'FATES: Invalid element in CheckIntergratedMassPools'

rgknox replied:
fixed

JessicaNeedham commented on main/ChecksBalancesMod.F90:
```diff
@@ -240,7 +243,134 @@ subroutine CheckLitterPools(currentSite,bc_in)
+            
+            !print*, ediag%err_liveveg!, ediag%err_liveveg/ibal%state_liveveg, ibal%state_liveveg, net_uptake, tot_litter_input
+
+            ! Perform the comparison between integrated flux and state
```
I think we discussed this but not sure where we landed on having this turned on or not. Could it be an option so a user could opt in to having their run crash if states and integrated fluxes start to diverge?

rgknox replied:
I added an issue here: https://github.com/NGEET/fates/issues/1264

JessicaNeedham reviewed (approved):
Thanks Ryan this looks great! My simulations with this branch merged also show very tiny divergence between integrated fluxes and stocks.

rgknox commented on biogeochem/EDPhysiologyMod.F90:
```diff
@@ -2878,9 +2888,9 @@ subroutine CWDInput( currentSite, currentPatch, litt, bc_in)
-       flux_diags%leaf_litter_input(pft) = &
-            flux_diags%leaf_litter_input(pft) +  &
-            leaf_m_turnover * currentCohort%n
+       elflux_diags%leaf_litter_input(pft) = &
```
renamed surf_fine_litter_input

rgknox commented on main/ChecksBalancesMod.F90:
```diff
@@ -240,7 +243,134 @@ subroutine CheckLitterPools(currentSite,bc_in)
     return
   end subroutine CheckLitterPools
 
+  ! ==================================================================
```
added descriptive comment

rgknox commented on main/FatesHistoryInterfaceMod.F90:
```diff
@@ -6433,12 +6422,6 @@ subroutine define_history_vars(this, initialize_variables)
                upfreq=group_nflx_simple, ivar=ivar, initialize=initialize_variables,              &
                index = ih_nefflux_si)
 
-          call this%set_history_var(vname='FATES_NDEMAND', units='kg m-2 s-1',      &
```
added them back in

rgknox commented:
@mpaiao and @JessicaNeedham I addressed your comments. @mpaiao I changed the name of that variable from leaf_* to surf_fine_* to be more general and include seed.  @JessicaNeedham I added back in those history variables and also created an issue about graceful failure.

rgknox commented:
All tests b4b except for the specific output variables that were changed in this PR with intent.
/glade/derecho/scratch/rgknox/tests_1017-183514de
